source/extensions/internal_redirect/previous_routes/previous_routes.cc

```diff
@@ -0,0 +1,52 @@
+#include "extensions/internal_redirect/previous_routes/previous_routes.h"
+
+#include "envoy/router/internal_redirect.h"
+#include "envoy/stream_info/filter_state.h"
+
+#include "absl/container/flat_hash_set.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace InternalRedirect {
+
+namespace {
+
+constexpr absl::string_view PreviousRoutesPredicateStateNamePrefix =
+    "envoy.internal_redirect.previous_routes_predicate_state";
+
+class PreviousRoutesPredicateState : public StreamInfo::FilterState::Object {
+public:
+  PreviousRoutesPredicateState() = default;
+  // Disallow copy so that we don't accidentally take a copy of the state
+  // through FilterState::getDataMutable, which will cause confusing bug that
+  // states are not updated in the original copy.
+  PreviousRoutesPredicateState(const PreviousRoutesPredicateState&) = delete;
+  PreviousRoutesPredicateState& operator=(const PreviousRoutesPredicateState&) = delete;
+
+  bool insertRouteIfNotPresent(absl::string_view route) {
+    return previous_routes_.insert(std::string(route)).second;
+  }
+
+private:
+  absl::flat_hash_set<std::string> previous_routes_;
+};
+
+} // namespace
+
+bool PreviousRoutesPredicate::acceptTargetRoute(StreamInfo::FilterState& filter_state,
```

Comment:
I think you're going to want stats and or trace logging for when you're trying to debug why some internal redirect was rejected rather than forwarded.   Maybe just trace log in the base class when a predicate check fails which one caused the failure, so you only have to implement once?

Reply:
Done. I ended up logging it directly in the router since I think ENVOY_STREAM_LOG is better. I did make the base InternalRedirectPredicate loggable so that implementations can do more logging if needed.